### PR TITLE
feat: expandable table

### PR DESCRIPTION
**Pull request by FezVrasta**
## PR checklist

- [x] I have followed the ["Committing and publishing"](https://github.com/quid/refraction/blob/master/CONTRIBUTING.md) guidelines;
- [x] I have added unit tests to cover my changes;
- [x] I updated the documentation and examples accordingly;

## Changes description

This PR ports the ExpandableTable to Refraction. The API is unchanged to help the migration.

This also fixes the internal tickets AF-598, AF-829, and AF-830

image

## Affected packages

- @quid/react-expandable-table

### CODE REVIEW

File: packages/react-expandable-table/src/index.js [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+  defaultSortOrder?: SortOrder,
+  // FIXME: this logic is weak, we should have a better way to control the sorting
+  // maybe just like we do for the openedRows logic, where we allow both a controlled
+  // and uncontrolled way to use the functionality. Keeping as is for now for backward compat
+  /**
+   * @deprecated When set to `true`, the first column will be automatically sorted by
+   * the first order value defined in `defaultSortOrder`, it will be replaced by a
+   * more flexible implementation in the future
+   */
+  defaultIsSorting?: boolean,
```
Comment (FezVrasta) [inline packages/react-expandable-table/src/index.js]: @gabrielmicko I'm not sure this is doing what the original component did, may you please review your implementation and describe what's the exact expected behavior? I couldn't figure it out from it.

If this is something we don't use I'd say we can directly remove it.

File: packages/react-expandable-table/src/styles.js [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+import * as React from 'react';
+import styled from '@emotion/styled/macro';
+import { GroupedVirtuoso } from 'react-virtuoso';
+import { themes, textStyles } from '@quid/theme';
+import { Icon, Text } from '@quid/react-core';
+import { Container } from '@quid/react-tooltip';
+import isPropValid from '@emotion/is-prop-valid';
+
+const alignToFlex = (align = 'left') =>
+  align === 'left' ? 'flex-start' : align === 'right' ? 'flex-end' : 'center';
```
Comment (pdw207) [inline packages/react-expandable-table/src/styles.js]: Hard to follow.

File: packages/react-expandable-table/src/styles.js [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+InfoIcon.defaultProps = {
+  theme: themes.dark,
+};
+
+const SortAsc = styled(props => <Icon name="sort_asc" {...props} />)``;
+const SortDesc = styled(props => <Icon name="sort_desc" {...props} />)``;
+const ACTIVE_ARROW = props =>
+  props.sort == null ? SortDesc : props.sort === 'desc' ? SortAsc : '.';
+const NEXT_ARROW = props =>
+  props.sort === 'desc' ? SortDesc : props.sort === 'asc' ? SortAsc : '.';
```
Comment (pdw207) [inline packages/react-expandable-table/src/styles.js]: Preference for clarity over simplicity.

File: packages/react-expandable-table/src/styles.js [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    ${ACTIVE_ARROW} {
+      color: ${props => props.theme.primary};
+    }
+  }
+`;
+SortIcon.defaultProps = {
+  theme: themes.dark,
+};
+
+export const TooltipContainer = styled(
```
Comment (pdw207) [inline packages/react-expandable-table/src/styles.js]: Are we checking the ref type? Not sure if flow raises an error here.

File: packages/react-expandable-table/src/utils.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+// @flow
+import { ASC, DESC, type Data } from './types';
+
+export const getSortedData = (
+  data: Array<Data>,
+  sortOrder: typeof ASC | typeof DESC | null,
```
Comment (pdw207) [inline packages/react-expandable-table/src/utils.js]: Not sure if it makes sense using null, what about using a default?

File: packages/react-expandable-table/src/styles.js (hunk 2 above)
Comment (FezVrasta) [reply packages/react-expandable-table/src/styles.js]: What's unclear? This is mapping the `align` property to the corresponding flex alignment value.

File: packages/react-expandable-table/src/styles.js (hunk 3 above)
Comment (FezVrasta) [reply packages/react-expandable-table/src/styles.js]: These are "dynamic" selectors, given the `sort` property they return the selector we want to target.

File: packages/react-expandable-table/src/utils.js (hunk 5 above)
Comment (FezVrasta) [reply packages/react-expandable-table/src/utils.js]: `null` is when no sorting is enabled, so the default data ordering is used.

File: packages/react-expandable-table/src/styles.js (hunk 4 above)
Comment (FezVrasta) [reply packages/react-expandable-table/src/styles.js]: I'm not sure I follow, I'm just passing down the `ref` to the `Container` component, because `@quid/react-tooltip` needs to access it.

File: packages/react-expandable-table/src/styles.js (hunk 3 above)
Comment (FezVrasta) [reply packages/react-expandable-table/src/styles.js]: I just realized I did invert the names, I fixed it. I guess it's clearer now :-)

**Comment by gabrielmicko** [comment]
Is it just me or hovering on tooltip is not working?

File: packages/react-expandable-table/src/index.js (hunk 1 above)
Comment (gabrielmicko) [reply packages/react-expandable-table/src/index.js]: It referred to if you want to sort on initial render or not. What we might want here is to say that which columns should be used for sorting on first render. That was missing from the other implementation as well.
```
  const [sorting, setSorting] = React.useState({
    key: defaultIsSorting ? columns[0].key : null, //-> maybe we should change defaultIsSorting to defaultSortBy so user can define which column should be sorted. What do you think?
    sort: defaultIsSorting ? defaultSortOrder[0] : null,
  });
```

File: packages/react-expandable-table/src/styles.js (hunk 2 above)
Comment (gabrielmicko) [reply packages/react-expandable-table/src/styles.js]: I agree here with @pdw207.

File: packages/react-expandable-table/src/index.js (hunk 1 above)
Comment (gabrielmicko) [reply packages/react-expandable-table/src/index.js]: maybe we should change defaultIsSorting to defaultSortBy so user can define which column should be sorted. What do you think?

Comment (FezVrasta) [reply packages/react-expandable-table/src/index.js]: I would rather remove this feature for now if we aren't using it, but I see we set it to `true` in our internal code so I wonder why...

**Comment by FezVrasta** [comment]
> Is it just me or hovering on tooltip is not working?

Fixed and added a test, thanks!

File: packages/react-expandable-table/src/styles.js (hunk 2 above)
Comment (FezVrasta) [reply packages/react-expandable-table/src/styles.js]: And I repeat my question 😛 What's unclear?

File: packages/react-expandable-table/src/index.js (hunk 1 above)
Comment (gabrielmicko) [reply packages/react-expandable-table/src/index.js]: Remove please

**Comment by gabrielmicko** [comment]
I will take a look again tmrw.

File: packages/react-expandable-table/src/styles.js (hunk 2 above)
Comment (pdw207) [reply packages/react-expandable-table/src/styles.js]: Multiple ternary operators in a row is hard to follow.

File: packages/react-expandable-table/src/styles.js (hunk 3 above)
Comment (pdw207) [reply packages/react-expandable-table/src/styles.js]: It like how elegant it is but it is hard for a reviewer to catch that mistake.

File: packages/react-expandable-table/src/styles.js (hunk 4 above)
Comment (pdw207) [reply packages/react-expandable-table/src/styles.js]: If req is not provided does it throw an error?

File: packages/react-expandable-table/src/utils.js (hunk 5 above)
Comment (pdw207) [reply packages/react-expandable-table/src/utils.js]: OK, got it. That makes sense.

File: packages/react-expandable-table/src/styles.js (hunk 4 above)
Comment (FezVrasta) [reply packages/react-expandable-table/src/styles.js]: No

File: packages/react-expandable-table/src/styles.js (hunk 2 above)
Comment (vampolo) [reply packages/react-expandable-table/src/styles.js]: I agree, a ternary operator seems good to make an if in a line. start nesting ternary operators is hard to follow. 
Please let's convert to if statements.
Let's optimize for readability than line counts.

File: packages/react-expandable-table/src/styles.js (hunk 4 above)
Comment (pdw207) [reply packages/react-expandable-table/src/styles.js]: If it's not required then ignore my comment.

File: packages/react-expandable-table/src/styles.js (hunk 3 above)
Comment (FezVrasta) [reply packages/react-expandable-table/src/styles.js]: I'm open to alternatives

**Comment by gabrielmicko** [comment]
Ellipsis & `bold: true` in the column row does not seem to be working.

**Comment by FezVrasta** [comment]
> Ellipsis & `bold: true` in the table row does not seem to be working.

Should ellipsis be applied to all the cells automatically?

**Comment by gabrielmicko** [comment]
> > Ellipsis & `bold: true` in the table row does not seem to be working.
> 
> Should ellipsis be applied to all the cells automatically?

We had it enabled yes, but it was a limitation because of VirtualList. I suggest to ask UX's opinion if we can break into more lines. If so maximum how many lines could that be?

**Comment by FezVrasta** [comment]
> We had it enabled yes, but it was a limitation because of VirtualList. I suggest to ask UX's opinion if we can break into more lines. If so maximum how many lines could that be?

Let's first ship this with feature parity, then we can improve it.

**Comment by gabrielmicko** [comment]
Would you please add some padding between ColumnCell's so if they are long they have some gap between them. I also miss having title on ColumnCell, so when the text is long and ellipsis is used at least title can tell the user whats hidden.

**Comment by gabrielmicko** [comment]
Looks good to me now. Thank you.
